Widen restrictive archive traversal, fixes #1925

Vogel612 opened the pull request:
----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

KengoTODA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/util/Archive.java:
```diff
@@ -70,9 +70,4 @@ private static String getExtension(String fileName) {
         return extension;
     }
-
-    public static boolean isLibraryFileName(String fileName) {
```
I think it's not so good to remove a public API by minor/patch update, could you revert this change and add `@Deprecated` annotation instead?

If you make a GitHub issue to remove this deprecated in the next major update, I will add it into the [milestone for v5](https://github.com/spotbugs/spotbugs/milestone/15).

KengoTODA reviewed (changes requested):
One request I left, please check. And if possible, try to [reproduce the issue you want to fix by a unit test](https://github.com/spotbugs/spotbugs/blob/31375986d170956c5a26f657752879b282601c1f/.github/CONTRIBUTING.md#before-reporting-a-problem-with-detectors).

Vogel612 commented:
@KengoTODA I have updated the public API as per your request. I am having a hard time finding the existing unit-tests relating to the ClassPathBuilder, though. Could you give me a hint where I should be looking?

Vogel612 commented:
@KengoTODA I added some unit-tests that rely on a handbuilt nested archive I added to the spotbugsTestCases. I really dislike this setup for the test case, but since I am not familiar enough with the codebase (and there was virtually no documentation on how something like this could even be attempted), that's the hack I went with.

Because these tests require acessing the engine configuration and I didn't want to mess with user preferences between test cases I added an overload to the SpotBugsRule used to integrate the Engine into UnitTests to support modifying the engine in code just before the execution of the actual analysis.

Vogel612 commented on test-harness-core/src/main/java/edu/umd/cs/findbugs/test/AnalysisRunner.java:
```diff
@@ -83,7 +84,7 @@ public AnalysisRunner addAuxClasspathEntry(Path path) {
 
     @Nonnull
-    public BugCollectionBugReporter run(Path... files) {
+    public BugCollectionBugReporter run(Consumer<IFindBugsEngine> engineCustomization, Path... files) {
```
The only use of that API is from within `SpotBugsRule`. I wouldn't define a testing interface adapter as public API. I had already adapted the API of `SpotBugsRule` to your suggestion, I don't think adding an overload here provides any benefit whatsoever.

KengoTODA reviewed (changes requested):
The code is LGTM. Thanks! Add an entry to `CHANGELOG.md` then I'll merge this PR for the coming release.

Vogel612 commented:
@KengoTODA Added a changelog entry and rebased onto the master branch. All commits have been squashed into a single one.

KengoTODA reviewed (approved):
Everything LGTM. Thanks a lot! :+1:
